packages/financial-templates-lib/src/price-feed/DominationFinancePriceFeed.js

```diff
@@ -0,0 +1,240 @@
+const { PriceFeedInterface } = require("./PriceFeedInterface");
+const { parseFixed } = require("@ethersproject/bignumber");
+
+// An implementation of PriceFeedInterface that uses DominationFinance's API to retrieve prices.
+class DominationFinancePriceFeed extends PriceFeedInterface {
+  /**
+   * @notice Constructs the DominationFinancePriceFeed.
+   * @param {Object} logger Winston module used to send logs.
+   * @param {Object} web3 Instance used for Web3 utilities and conversions.
+   * @param {String} pair Representation of the pair the price feed is tracking.
+   *    The string should be the representation used by the DomFi API to identify this pair.
+   * @param {Integer} lookback How far in the past the historical prices will be available using getHistoricalPrice.
+   * @param {Object} networker Used to send the API requests.
+   * @param {Function} getTime Returns the Unix timestamp in seconds.
+   * @param {Integer} minTimeBetweenUpdates Min number of seconds between updates. If update() is called again before
+   *      this number of seconds has passed, it will be a no-op.
+   * @param {Number} decimals Number of decimals to use to convert price to wei.
```

Comment:
This is great! Note it looks like you build your `DominationFinancePriceFeed` by mimicking the `CryptoWatchPriceFeed`, which is the perfect way to do this seeing it's got the right layout and structure.

There are a few small nits/changes I'd like you to make. After you copied  `CryptoWatchPriceFeed` we've made some pretty meaningful changes to how we deal with decimals within our feeds. In particular, feeds can be scaled to non-18 decimals to accommodate contracts with non-18 decimal collateral types. Note that the `CryptowatchPriceFeed` _could_ previously do this but we did make some changes to the flow. you can check it out here: https://github.com/UMAprotocol/protocol/pull/2276

To make your PR compatible with how we now deal with decimals could you please make the following changes:

- rename `decimals` to `priceFeedDecimals` throughout this PR.
- rename `convertDecimals` to `convertPriceFeedDecimals` throughout this PR.
- add a new method `priceFeedDecimals` that simply returns `return this.priceFeedDecimals`.

After you've made these changes you should be up to date with our latest structure. Because your feeds are 18 decimal scaled I don't think you really need to change anything else.

Reply:
Just to be on the same page with the "number of decimals" logic being used here -- so `priceFeedDecimals` is solely there to allow the client of the `*PriceFeed` classes the ability to shift the number of decimal places the BigNumber returns, correct? This doesn't have anything to do with the UMIP rounding rules, right? (At the moment, the API will perform the appropriate rounding according to the UMIP and returns those values as a string for simplicity)

I'll get those renamed and add the new `priceFeedDecimals` method 👍🏾

Reply:
`priceFeedDecimals` defines the scaling of the feed itself and can be parameterized by the config passed into the feed. The point of it is to scale feeds in the _same way the DVM would return_ for a given vote. For older EMP products (and some that are currently on mainnet) that use non-18 decimal collateral types we would _scale the price feed_ according to the decimals of the collateral type. For example, contracts collateralized in BTC would have:
- 18 decimal synthetic
- 8 decimal collateral type
- 8 decimal price feed

For these non-18 decimal collateral types, we need to change how the price feed is called to match the collateral.

In https://github.com/UMAprotocol/protocol/pull/2276 we are changing this pattern to accommodate our new contracts that are currently in an audit by OZ. In the future, we will match the decimals of the synthetic and collateral and keep all feeds at 18 decimals. The above BTC example would then look like this:
- 8 decimal synthetic
- 8 decimal collateral type
- 18 decimal price feed

`priceFeedDecimals` within a price feed enables the feed to accommodate _either_ kind of decimals so that the bots remain backward compatible. The actual computation of where this scaling is used can be found within the EMP client, for example, which can be found here: https://github.com/UMAprotocol/protocol/blob/133c4a6e094e7c5dd35673b41a1ee13fd676b86a/packages/financial-templates-lib/src/clients/ExpiringMultiPartyClient.js#L244

In the case of your product you want the price feed to be scaled to 18 decimals because a) your collateral type is Dai and b) you deployed with the old contracts that expect 18 decimal feeds.

Does that make sense?

Reply:
@chrismaree  Thanks, that makes a lot more sense with how the decimals are working from a high level between all the different components 👍🏾